server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/ImapKeywordsConsistencyContract.scala

```diff
@@ -0,0 +1,614 @@
+
+    // When the user asks for message list with flag "$Flagged"
+    val ids = listMessageIdsBykeyword("$Flagged").asScala.toList
+
+    // Then the message list has size 1
+    // And the message list contains "m1"
+    assertThat(ids.size).isEqualTo(1)
+    assertThat(ids).isEqualTo(List(messageId.serialize))
+  }
+
+  @Test
+  def emailQueryInSpecificMailboxShouldReturnMessageIdWhenMatching(server: GuiceJamesServer): Unit = {
+    // Given the user has a message "m1" in "inbox" mailbox with subject "My awesome subject", content "This is the content"
+    val messageId = appendMessageToInbox(server)
+
+    // And user copies "m1" from mailbox "inbox" to mailbox "archive"
+    copyMessageFromInboxToArchive(server, messageId)
+    awaitAtMostOneMinute.until(() => listMessageIdsArchive(server).size == 1)
+
+    // And the user has an open IMAP connection with mailbox "archive" selected
+    imapClient.connect(LOCALHOST_IP, server.getProbe(classOf[ImapGuiceProbe]).getImapPort)
+      .login(BOB, BOB_PASSWORD)
+      .select(DefaultMailboxes.ARCHIVE)
+
+    // And the user set flags via IMAP to "(\Flagged)" for all messages in mailbox "archive"
+    imapClient.setFlagsForAllMessagesInMailbox("\\Flagged")
+
+    // When user asks for message list in mailbox "archive" with flag "$Flagged"
+    val ids = listMessageIdsByMailboxAndKeyword(server, DefaultMailboxes.ARCHIVE, "$Flagged").asScala.toList
+
+    // Then the message list has size 1
+    // And the message list contains "m1"
+    assertThat(ids.size).isEqualTo(1)
+    assertThat(ids).isEqualTo(List(messageId.serialize))
+  }
+
+  @Test
+  def emailQueryInSpecificMailboxShouldSkipMessageIdWhenNotMatching(server: GuiceJamesServer): Unit = {
+    // Given the user has a message "m1" in "inbox" mailbox with subject "My awesome subject", content "This is the content"
+    val messageId = appendMessageToInbox(server)
+
+    // And user copies "m1" from mailbox "inbox" to mailbox "archive"
+    copyMessageFromInboxToArchive(server, messageId)
+    awaitAtMostOneMinute.until(() => listMessageIdsArchive(server).size == 1)
+
+    // And the user has an open IMAP connection with mailbox "archive" selected
+    imapClient.connect(LOCALHOST_IP, server.getProbe(classOf[ImapGuiceProbe]).getImapPort)
+      .login(BOB, BOB_PASSWORD)
+      .select(DefaultMailboxes.ARCHIVE)
+
+    // And the user set flags via IMAP to "(\Flagged)" for all messages in mailbox "archive"
+    imapClient.setFlagsForAllMessagesInMailbox("\\Flagged")
+
+    // When user asks for message list in mailbox "inbox" with flag "$Flagged"
+    val ids = listMessageIdsByMailboxAndKeyword(server, DefaultMailboxes.INBOX, "$Flagged").asScala.toList
+
+    // Then the message list is empty
+    assertThat(ids.size).isEqualTo(0)
+  }
+
+  @Disabled("TODO: Fix!")
+  @Test
+  def emailSetShouldSucceedToSolveKeywordsConflictsIntroducedViaImapUponFlagsAddition(server: GuiceJamesServer): Unit = {
```

Comment:
That one is a tricky one, tried to solve it quickly but every time it created problems elsewhere... 

I think it's better to leave it disabled for now, and fix it in a following PR

Reply:
Created a task for it => https://github.com/linagora/james-project/issues/5158

This PR is ready for review

Reply:
Please create an ASF task and put ASF ticket number into the @Disabled description

Reply:
JAMES-4026